File: inventory-service/src/main/java/com/catius/inventory/repository/InventoryRepository.java
package com.catius.inventory.repository;

import com.catius.inventory.domain.Inventory;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

public interface InventoryRepository extends JpaRepository<Inventory, Long> {

    @Modifying
    @Query(value = """
            UPDATE inventory
               SET quantity = quantity - :quantity
             WHERE product_id = :productId
               AND quantity >= :quantity
            """, nativeQuery = true)
    int decrementIfSufficient(@Param("productId") long productId,
                              @Param("quantity") int quantity);

    @Modifying
    @Query(value = """
            UPDATE inventory
               SET quantity = quantity + :quantity
             WHERE product_id = :productId
            """, nativeQuery = true)
    int increment(@Param("productId") long productId,
                  @Param("quantity") int quantity);
}

File: inventory-service/src/main/kotlin/com/catius/inventory/domain/port/InventoryRepository.kt
package com.catius.inventory.domain.port

import com.catius.inventory.domain.Inventory

interface InventoryRepository {
    fun save(inventory: Inventory): Inventory

    /** 락 없이 단순 조회 (find usecase 용). */
    fun findByProductId(productId: Long): Inventory?

    /** 비관적 쓰기 락으로 조회. 차감/복원 흐름에서 사용. */
    fun findByProductIdForUpdate(productId: Long): Inventory?
}
